{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "10526e03-9273-43ba-b907-05198ed2cffa",
   "metadata": {},
   "source": [
    "# This notebook must be run with the ABC_Download conda environment within the ABC.sif singularity container"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "482e0601-6753-4c3f-9455-b6f2bdb820d3",
   "metadata": {},
   "source": [
    "This code, and the following two notebooks, are largely inspired by the Allen Brain Cell Atlas's\n",
    "MERFISH atlas. The code used was sourced and adapted from this github: https://github.com/ZhuangLab/whole_mouse_brain_MERFISH_atlas_scripts_2023/blob/main/scripts/integrate_MERFISH_with_scRNA-seq/create_integration_workspace.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a5e41a1-722a-489f-9fd9-76c3cb81d1bb",
   "metadata": {},
   "source": [
    "Downloading the Allen Brain Cell Atlas: 20230521 release https://data.nemoarchive.org/other/grant/aibs_internal/zeng/transcriptome/scell/10x_v3/mouse/processed/counts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "25582e4f-b757-4b27-80fd-fe488471e0a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import igraph\n",
    "import networkx as nx\n",
    "import metis\n",
    "import ALLCools\n",
    "from ALLCools.integration.seurat_class import SeuratIntegration\n",
    "from scipy.sparse import csr_matrix\n",
    "import scipy.sparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d59360cf-ac31-42ee-b214-f9eb44613543",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_path = '/hpc/projects/group.quake/doug/references/ABC/'\n",
    "adata_file = f'{base_path}whole/AIT17.0.rawcount_logCPM_10Xv3/AIT17.0.rawcount_logCPM_10Xv3.h5ad'\n",
    "annotation_file = f'{base_path}AIT17.0.cl.df.v6_lock_230504/AIT17.0.cl.df.v6_lock_230504.tsv'\n",
    "\n",
    "annotation_df = pd.read_csv(annotation_file, sep='\\t')\n",
    "annotation_df['cl'] = annotation_df['cl'].astype(str)\n",
    "subclass_map = {t[0] : t[1] for t in annotation_df[['cl', 'subclass_label']].itertuples(index=False)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "aba80a1d-9193-48fb-a44a-ee0e5ad6b526",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_path = '/hpc/projects/group.quake/doug/references/ABC/whole/'\n",
    "adata = sc.read_h5ad(adata_file)\n",
    "adata.obs['subclass_label'] = adata.obs['cl'].map(subclass_map)\n",
    "\n",
    "# Filter out cells without subclass annotations or annotated as low quality\n",
    "mask_not_na = ~adata.obs['subclass_label'].isna()\n",
    "mask_not_lq = adata.obs['subclass_label'] != 'LQ'\n",
    "adata = adata[mask_not_na & mask_not_lq].copy()\n",
    "gc.collect()\n",
    "\n",
    "adata_raw = anndata.AnnData(X=adata.layers['rawcount'], obs=adata.obs, var=adata.var)\n",
    "adata_raw.write_h5ad(f'{out_path}AIT17_10Xv3.h5ad', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "5039226c-80ec-4f01-a258-acf479ae01bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Workspace creation\n",
    "def partition_subclasses(adata, n_partitions):\n",
    "    '''Assign integration partitions as the adata.obs['integration_partion'] column.'''\n",
    "    adata.obs['integration_partition'] = 'p0'\n",
    "    \n",
    "    # Get a sparse matrix of that counts edges between subclasses\n",
    "    ones = adata.obsp['distances'].copy()\n",
    "    ones.data = np.ones(len(ones.data))\n",
    "\n",
    "    g_neighobrs = sc._utils.get_igraph_from_adjacency(ones, directed=True)\n",
    "    vc = igraph.VertexClustering(g_neighobrs, membership=adata.obs['subclass_label'].cat.codes.values)\n",
    "    cluster_graph = vc.cluster_graph(combine_edges='sum')\n",
    "    #cluster_mtx = sc._utils.get_sparse_from_igraph(vc.cluster_graph(combine_edges='sum'), weight_attr='weight')\n",
    "    edges = cluster_graph.get_edgelist()\n",
    "    weights = [e['weight'] for e in cluster_graph.es]\n",
    "    num_clusters = len(vc)\n",
    "    row_indices, col_indices = zip(*edges)\n",
    "    cluster_mtx = csr_matrix((weights, (row_indices, col_indices)), shape=(num_clusters, num_clusters))\n",
    "\n",
    "    \n",
    "    # Partition the cluster level graph\n",
    "    G_cluster = nx.from_scipy_sparse_array(cluster_mtx)\n",
    "    for i in adata.obs['subclass_label'].cat.codes.values:\n",
    "        G_cluster.nodes[i]['weight'] = np.sum(adata.obs['subclass_label'] \n",
    "                                              == adata.obs['subclass_label'].cat.categories[i] )\n",
    "    G_cluster.graph['node_weight_attr'] = 'weight'\n",
    "    \n",
    "    (cut, parts) = metis.part_graph(G_cluster, n_partitions, recursive=False,\n",
    "                                   tpwgts=[1 / n_partitions] * n_partitions) \n",
    "    \n",
    "    # Assign the partitions\n",
    "    for i in range(len(parts)):\n",
    "        mask = (adata.obs['subclass_label'] == adata.obs['subclass_label'].cat.categories[i])\n",
    "        adata.obs.loc[mask, 'integration_partition'] = 'p' + str(parts[i])\n",
    "    \n",
    "    adata.obs['integration_partition'] = adata.obs['integration_partition'].astype('category')\n",
    "    \n",
    "def get_cluster_mean_expression_matrix(adata, cluster_column):\n",
    "    '''Get a dataframe of mean gene expression of each cluster.'''\n",
    "    if scipy.sparse.issparse(adata.X):\n",
    "        X = adata.X.toarray()\n",
    "    else:\n",
    "        X = adata.X\n",
    "   \n",
    "    cell_exp_mtx = pd.DataFrame(X, index=adata.obs[cluster_column], columns=adata.var.index)    \n",
    "    return cell_exp_mtx.groupby(by=cluster_column).mean()\n",
    "\n",
    "def calculate_correlation_matrix(adata, cluster_column):\n",
    "    '''Calculate gene expression correlations between clusters.'''\n",
    "    cluster_mean_exp = get_cluster_mean_expression_matrix(adata, cluster_column)\n",
    "    \n",
    "    # Initialzie the correlation matrix\n",
    "    cluster_ids = np.array(cluster_mean_exp.index)\n",
    "    N_clusters = len(cluster_ids)\n",
    "    correlation_mtx = pd.DataFrame(np.ones((N_clusters, N_clusters)), index=cluster_ids, columns=cluster_ids)\n",
    "    \n",
    "    # Fill the correlation matrix\n",
    "    for i in range(N_clusters):\n",
    "        cluster_id1 = cluster_ids[i]\n",
    "        mean_exps1 = np.array(cluster_mean_exp.loc[cluster_id1])\n",
    "        \n",
    "        for j in range(i + 1, N_clusters):\n",
    "            cluster_id2 = cluster_ids[j]\n",
    "            mean_exps2 = np.array(cluster_mean_exp.loc[cluster_id2])\n",
    "            \n",
    "            r, p = scipy.stats.pearsonr(mean_exps1, mean_exps2)\n",
    "\n",
    "            correlation_mtx.loc[cluster_id1, cluster_id2] = r\n",
    "            correlation_mtx.loc[cluster_id2, cluster_id1] = r\n",
    "            \n",
    "    return correlation_mtx\n",
    "\n",
    "def merge_clusters(obs_df, cluster_col, clusters_to_merge):\n",
    "    merged_cluster_id = sorted(clusters_to_merge)[0]\n",
    "    obs_df[cluster_col][obs_df[cluster_col].isin(clusters_to_merge)] = merged_cluster_id\n",
    "    obs_df[cluster_col] = obs_df[cluster_col].cat.remove_unused_categories()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "15e2721f-fc31-4e4d-8163-777d8431b42f",
   "metadata": {},
   "outputs": [],
   "source": [
    "workspace_path = f'{out_path}integration_workspace'\n",
    "input_seq_file = f'{out_path}AIT17_10Xv3.h5ad'\n",
    "\n",
    "# Make the top level directories\n",
    "os.makedirs(workspace_path, exist_ok=True)\n",
    "partition_path = os.path.join(workspace_path, 'partitions')\n",
    "os.makedirs(partition_path, exist_ok=True)\n",
    "os.makedirs(os.path.join(workspace_path, 'gene_expression_imputation'), exist_ok=True)\n",
    "\n",
    "#%%time\n",
    "# Load the sequencing data\n",
    "adata_seq = sc.read_h5ad(input_seq_file)\n",
    "\n",
    "adata_seq.obs['subclass_label'] = adata_seq.obs['subclass_label'].astype('category')\n",
    "adata_seq.obs['cl'] = adata_seq.obs['cl'].astype('category')\n",
    "\n",
    "# Load the merfish data\n",
    "adata_merfish = sc.read_h5ad('FINAL_ANNOTATION.h5ad')\n",
    "\n",
    "# Rename the genes with different synonyms names\n",
    "adata_merfish.var.rename(index={\n",
    "    'AC102910.1': 'Gm30564', \n",
    "    'BC030499': 'Rskr',\n",
    "    'Ctgf': 'Ccn2',\n",
    "    'Fam196b': 'Insyn2b',\n",
    "    'Fam19a1': 'Tafa1',\n",
    "    'Fam19a2': 'Tafa2',\n",
    "    'Fam19a4': 'Tafa4',\n",
    "    'Fam46a': 'Tent5a',\n",
    "    'Fam84b': 'Lratd2',\n",
    "    'Nov': 'Ccn3',\n",
    "    'Wisp1': 'Ccn4',\n",
    "}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1d4e3bfd-7160-46a0-a705-6c6f68fa4e29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Consider the common genes\n",
    "common_genes = np.array(adata_seq.var_names.intersection(adata_merfish.var_names))\n",
    "adata_merfish = adata_merfish[:, adata_merfish.var.index.isin(common_genes)]\n",
    "adata_merfish.write_h5ad(os.path.join(workspace_path, 'adata_merfish.h5ad'), compression='gzip')\n",
    "\n",
    "adata_seq._inplace_subset_var(common_genes)\n",
    "\n",
    "# Remove the cells without cluster or subclass labels\n",
    "adata_seq = adata_seq[~adata_seq.obs['cl'].isna()]\n",
    "adata_seq = adata_seq[~adata_seq.obs['subclass_label'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "561c9b1a-e980-4691-8ca7-50dbc4b6ab6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/doug.henze/.conda/envs/ABC_Download/lib/python3.9/site-packages/scanpy/preprocessing/_normalization.py:207: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n",
      "/home/doug.henze/.conda/envs/ABC_Download/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/tmp/ipykernel_101497/2955554804.py:46: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  return cell_exp_mtx.groupby(by=cluster_column).mean()\n"
     ]
    }
   ],
   "source": [
    "#%%time\n",
    "# Define the integration partitions using the sequencing data\n",
    "n_partitions = 80\n",
    "\n",
    "sc.pp.normalize_total(adata_seq, target_sum=1000)\n",
    "sc.pp.log1p(adata_seq)\n",
    "sc.pp.scale(adata_seq)\n",
    "\n",
    "n_pcs=100\n",
    "sc.tl.pca(adata_seq, svd_solver='arpack', n_comps=n_pcs)\n",
    "sc.pp.neighbors(adata_seq, use_rep='X_pca', n_neighbors=15, n_pcs=n_pcs)\n",
    "\n",
    "partition_subclasses(adata_seq, n_partitions)\n",
    "\n",
    "#Merge partitions with too few cells\n",
    "partition_corr_df = calculate_correlation_matrix(adata_seq, 'integration_partition')\n",
    "p_names, p_n_cells = np.unique(adata_seq.obs['integration_partition'], return_counts=True)\n",
    "p_smalls = p_names[p_n_cells < 10000]\n",
    "p_larges = p_names[p_n_cells >= 10000]\n",
    "partition_corr_df = partition_corr_df.loc[p_smalls, p_larges]\n",
    "\n",
    "for p1 in p_smalls:\n",
    "\n",
    "    p2 = partition_corr_df.columns[np.argmax(partition_corr_df.loc[p1])]\n",
    "    \n",
    "    print(f'Merge partitions {p1} and {p2}')\n",
    "    merge_clusters(adata_seq.obs, 'integration_partition', [p1, p2])\n",
    "    \n",
    "seq_annotation_df = adata_seq.obs.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "174b32ed-05a0-4fc8-b397-0a9d430a263b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%time\n",
    "# Reload the sequencing data and assign the partitions\n",
    "adata_seq = sc.read_h5ad(input_seq_file)\n",
    "adata_seq.obs['integration_partition'] = seq_annotation_df['integration_partition']\n",
    "\n",
    "adata_seq.obs['subclass_label'] = adata_seq.obs['subclass_label'].astype('category')\n",
    "adata_seq.obs['cl'] = adata_seq.obs['cl'].astype('category')\n",
    "\n",
    "adata_seq = adata_seq[~adata_seq.obs['cl'].isna()]\n",
    "adata_seq = adata_seq[~adata_seq.obs['subclass_label'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3221802f-bc8d-4a58-a054-fb91caaedc83",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_annotation_df.to_csv('seq_annotation_df.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f1afbf25-e1c5-4464-84b8-9c9b9063ecf5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Partition, N_cells, subclasses\n",
      "p0, 28842, ['LSX Nkx2-1 Gaba', 'LSX Prdm12 do Gaba', 'MEA-BST Lhx6 Nfib Gaba', 'PAG Pou4f2 Glut', 'SUB-ProS Glut']\n",
      "p1, 30035, ['IPN-IF-RL Foxa1 Glut', 'LSX Sall3 Lmo1 Gaba', 'MEA Otp Glut', 'MEA-BST Lhx6 Nr2e1 Gaba', 'MEA-BST Lhx6 Sp9 Gaba', 'MPO-LPO Lhx8 Gaba', 'MPO-MPN Glut', 'PAG-SC Pou4f2 Mesi2 Glut', 'SBPV-PVi Six3 Gaba', 'TMd Foxd2 Gaba']\n",
      "p10, 30080, ['AHN-SBPV Gaba', 'ARH-PVp Tbx3 Glut', 'DMH Hmx2 Gaba', 'NDB-SI-ant Prdm12 Gaba', 'NTS Dbh Glut', 'PH-SUM Foxa1 Glut', 'PPN-CUN-PAG Pax5 Sox21 Gaba', 'PPN-CUN-PCG Otp En1 Gaba', 'PVH Six3 Prox1 Gaba', 'PVH-LHA Barhl2 Glut', 'PVH-PVHd-SO Otp Glut', 'PVHd-DMH Lhx6 Gaba', 'PVR Six3 Gaba', 'RN-MRN-CLI En1 Sox2 Gaba', 'SCs Pax7 Nfia Gaba']\n",
      "p11, 29841, ['COAa-PAA-MEA Barhl2 Glut', 'PRC-PAG Pax6 Glut', 'PVpo-VMPO-SFO Tbr1 Glut', 'SI-MA-LPO Skor1 Glut', 'SI-MPO-LPO Lhx8 Gaba', 'SPA-SPFm-SPFp-POL-PIL-PoT Glut', 'TU-ARH Otp Six6 Gaba']\n",
      "p12, 29751, ['CUN Evx2 Lhx2 Glut', 'CUN-PPN Evx2 Meis2 Glut', 'DMH-LHA Gsx1 Gaba', 'DTN-PCG-IPN Otp Pax3 Gaba', 'LDT-PCG Vsx2 Lhx4 Glut', 'PAL-STR Gaba-Chol', 'PH-ant Otp Glut', 'SC Lef1 Otx2 Gaba', 'SNc-VTA-RAmb Foxa1 Dopa']\n",
      "p13, 30149, ['AHN Onecut3 Gaba', 'BST-MPN Lhx6 Zfhx3 Gaba', 'CBN Glut', 'CS-PDTg Pax6 Gaba', 'DTN-LDT-PCG St18 Gaba', 'IPN-LDT Vsx2 Nkx6-1 Glut', 'IRN Dmbx1 Pax2 Gly-Gaba', 'IRN Vip Glut', 'LDT Fgf7 Gaba', 'LDT-PCG Gata3 Irx5 Gaba', 'MARN-PPY Ngfr Gly-Gaba', 'MDRNd Bves Glut', 'MRN-LDT-DTN Pax8 Gaba', 'MRN-VTN-DTN Pax5 Cdh23 Gaba', 'MS-SF Bsx Glut', 'MV Pax6 Gly-Gaba', 'MV-SPIV Phox2b Ebf3 Lbx1 Glut', 'NLL Gata3 Gly-Gaba', 'PAG Dmrta2 Glut', 'PAG Nkx2-2 Nr2f2 Gaba', 'PB Pax5 Glut', 'PCG-RPO-CS-PRNr Tox Gaba', 'PDTg Otp Olig3 Gaba', 'PDTg Otp Shroom3 Gaba', 'PGRN-PARN-MDRN Hoxb5 Glut', 'PH-an Pitx2 Glut', 'POR Gata3 Gly-Gaba', 'PRP Gata3 Slc6a5 Gly-Gaba', 'PVa-ARH Six3 Dopa-Gaba', 'RO-RPA Pkd2l1 Gaba', 'RPA Pax6 Hoxb5 Gly-Gaba', 'RPO-NI Vsx2 Nr4a2 Glut', 'SC Tnnt1 Gli3 Gaba', 'SNr Six3 Gaba', 'SNr-VTA-PPN Pax5 Npas1 Gaba', 'SPVO Mafa Meis2 Glut']\n",
      "p14, 30188, ['ADP-MPO Trp73 Glut', 'BST Tac2 Gaba', 'CEA-BST Crh Gaba', 'DMH Hmx2 Glut', 'DMX Tbx20 Chol', 'GRN Prox1 Brs3 Gly-Gaba', 'HY Gnrh1 Glut', 'IC Six3 En2 Gaba', 'LDT Vsx2 Nkx6-1 Nfib Glut', 'LGv Otx2 Gaba', 'LGv-SPFp-SPFm Gata3 Gaba', 'LHA Pmch Glut', 'MB-MY Glut-Sero', 'MB/HB Lhx1 Glut', 'MEA-BST Gal Avp Gaba', 'MG-POL-SGN Glut', 'PB-NTS Phox2b Ebf3 Lmx1b Glut', 'PH Pitx2 Glut', 'PMv-TMv Tbx3 Hist-Gaba', 'Pineal Crx Glut', 'SCop Pou4f2 Neurod2 Glut', 'SI-MA-NDB Six3 Meis2 Gaba', 'VMH Fezf1 Glut']\n",
      "p15, 29304, ['MOB Eomes Ms4a15 Glut', 'NDB-SI-MA-STRv Lhx8 Gaba', 'NTS Phox2b Glut', 'PAG Pou4f1 Ebf2 Glut', 'PAG-SC Pou4f1 Glut', 'PRC-PAG Tcf7l2 Glut']\n",
      "p16, 29745, ['CBX Golgi Gly-Gaba', 'DMH-LHA Vgll2 Glut', 'PAG Pou4f1 Glut', 'PAG Pou4f3 Glut', 'PAG Tfap2b Glut', 'PARN-MDRNd Gbx2 Gly-Gaba', 'PM Pitx2 Glut', 'PRP-NI-PRNc-GRN Otp Glut', 'RE-Xi Nox4 Glut', 'SCdg Tfap2b Glut', 'SCig-ve Pitx2 Glut', 'ZI Pax6 Gaba']\n",
      "p17, 30215, ['AHN-RCH-TU-SBPV Otp Glut', 'APN C1ql4 Glut', 'DCO Il22 Gly-Gaba', 'GPe Sox6 Cyp26b1 Gaba', 'IA Mgp Gaba', 'MPT-NOT-PPT Mecom Gaba', 'PG-TRN-LRN Fat2 Glut', 'PM-LHA Foxb1 Glut', 'PRP-HB Otp Gly-Gaba', 'RN-PAG Nkx2-2 Otx1 Gaba', 'SC-PAG Lef1 Emx2 Gaba', 'SCs Lef1 Gli3 Gaba']\n",
      "p18, 30076, ['AD Serpinb7 Glut', 'APN C1ql2 Glut', 'CU-ECU-SPVI Foxb1 Glut', 'GPi Skor1 Gaba-Glut', 'IO Glut', 'LH Pou4f1 Sox1 Glut', 'MA3 Foxd2 Glut', 'MV-SPIV Glut', 'MV-SPIV-PRP Dmbx1 Gly-Gaba', 'MY Calcb Chol', 'ND-PCG Gata3 Lhx1 Gaba', 'PAG-MRN Pou3f1 Glut', 'PB Lmx1a Glut', 'PB-CUN Ppp1r1c Glut', 'PCG-PRN-GRN Vsx2 Nkx6-1 Glut', 'POR Spp1 Gly-Gaba', 'PSV Glut', 'PVa-AHN-LHA Otp Glut', 'RN-MRN-PAG Foxa2 Gaba', 'SCig Tfap2b Chrnb3 Glut', 'SPVI-SPVC Tlx3 Ebf3 Glut', 'STN-PSTN Pitx2 Glut', 'VCO-SOCl Mafa Meis2 Glut']\n",
      "p19, 30063, ['CBN Dmbx1 Gaba', 'LHA-MEA Otp Glut', 'MV Nkx6-1 Gly-Gaba', 'MY Lhx1 Gly-Gaba', 'PB-an-do Evx2 Glut', 'PF Fzd5 Glut', 'PH-LHA Foxb1 Glut', 'PRT Tcf7l2 Gaba', 'RT ZI Gnb3 Gaba', 'SCs Dmbx1 Gaba']\n",
      "p2, 29993, ['CEA-AAA-BST Ebf1 Gaba', 'LSX Prdm12 ve Gaba', 'NTS-DMX-PARN Neurod2 Gly-Gaba', 'SCig Foxb1 Glut']\n",
      "p20, 36488, ['MSN D2 Gaba']\n",
      "p21, 20003, ['L5 IT CTX Glut']\n",
      "p22, 33672, ['MOB-in Frmd7 Gaba']\n",
      "p23, 25457, ['BST-po Glut', 'L6 IT CTX Glut', 'PAG Ucn Glut', 'PAG-MRN Rln3 Gaba', 'PRNr Otp Nfib Glut']\n",
      "p24, 35778, ['MSN D1 Gaba']\n",
      "p25, 30141, ['AV Col27a1 Glut', 'L2/3 IT ENTl-PIR Glut', 'L5 NP CTX Glut', 'L6b CTX Glut', 'L6b EPd Glut']\n",
      "p26, 30219, ['L2/3 IT PPP Glut', 'L2/3 IT RSP Glut', 'L5 ET CTX Glut', 'LA-BLA-BMA-PA Glut']\n",
      "p27, 28688, ['MOB-out Frmd7 Gaba', 'MSN D1 Sema5a Gaba', 'OT D3 Folh1 Gaba', 'Sncg Gaba']\n",
      "p28, 30096, ['Lamp5 Gaba', 'Sst Gaba']\n",
      "p29, 29502, ['Lamp5 Lhx6 Gaba', 'MOB-STR-CTX Inh IMN', 'Pvalb Gaba', 'Pvalb chandelier Gaba', 'RHP-COA Ndnf Gaba', 'STR Prox1 Lhx6 Gaba', 'Sst Chodl Gaba']\n",
      "p3, 30190, ['ACB-BST-FS D1 Gaba', 'ARH-PVp Tbx3 Gaba', 'DG-PIR Ex IMN', 'DMH Nkx2-4 Glut', 'DMH Prdm13 Gaba', 'IPN Otp Crisp1 Gaba', 'LSX Sall3 Pax6 Gaba', 'MEA-BST Otp Glut', 'MOB Dopa-Gaba', 'MOB Meis2 Gaba', 'MOB Trdn Gaba', 'MOB-mi Frmd7 Gaba', 'RPO-CS-NI Meis2 Gaba', 'SC Otx2 Gcnt4 Gaba', 'SCH Gaba', 'SI-MA-NDB Ebf1 Bnc2 Gaba', 'SPVC Ccdc172 Glut', 'SPVC Nmu Glut', 'STRv Lhx8 Gaba']\n",
      "p33, 107182, ['OPC NN']\n",
      "p36, 52648, ['L6 CT CTX Glut']\n",
      "p37, 31675, ['CBX MLI Megf11 Gaba']\n",
      "p38, 49480, ['L2/3 IT CTX Glut']\n",
      "p39, 26503, ['ECU-CU Slc6a2 Glut', 'MRN-PAG Nkx6-1 Glut', 'NTS Crp Glut', 'PVHd Gsc Gaba', 'TH Prkcd Grin2c Glut']\n",
      "p4, 30178, ['CEA-AAA-BST Six3 Cyp26b1 Gaba', 'LSX Otx2 Gaba', 'PAS-MV Ebf2 Gly-Gaba', 'PVpo-VMPO-MPN Hmx2 Gaba', 'SCm-PAG Cdh23 Gaba', 'SPVI-SPVC Sall3 Gly-Gaba', 'VMH Nr5a1 Glut']\n",
      "p44, 145025, ['Astro-TE NN']\n",
      "p45, 28577, ['ABC NN', 'Astro-CB NN', 'Astro-OLF NN', 'Astroependymal NN', 'Bergmann NN', 'CHOR NN', 'DC NN', 'Ependymal NN', 'Hypendymal NN', 'Lymphoid NN', 'Monocytes NN', 'OEG NN', 'SMC NN', 'Tanycyte NN']\n",
      "p46, 28865, ['BAM NN', 'Peri NN']\n",
      "p48, 93100, ['Endo NN', 'VLMC NN']\n",
      "p53, 138489, ['CB Granule Glut', 'CB PLI Gly-Gaba', 'CBX Purkinje Gaba']\n",
      "p55, 29364, ['CA1-ProS Glut', 'CA2-FC-IG Glut', 'COAp Grxcr2 Glut', 'COAp-TR-ENTmv Glut', 'CT SUB Glut', 'L2 IT ENT-po Glut', 'L2 IT PPP-APr Glut', 'L2/3 IT ENT Glut', 'L4 RSP-ACA Glut', 'L5 PPP Glut', 'L5/6 IT TPE-ENT Glut', 'NLOT Rho Glut', 'NP PPP Glut', 'NP SUB Glut']\n",
      "p56, 29554, ['DG Glut', 'IT EP-CLA Glut', 'L6b/CT ENT Glut']\n",
      "p59, 93627, ['CA3 Glut', 'DCO UBC Glut', 'L4/5 IT CTX Glut']\n",
      "p6, 34387, ['IC Tfap2d Maf Glut', 'STR-PAL Chst9 Gaba']\n",
      "p69, 388671, ['Oligo NN', 'Vip Gaba']\n",
      "p7, 30111, ['CM-IAD-CL-PCN Glut', 'MEA-COA-BMA Ccdc42 Glut', 'MM Foxb1 Glut', 'PAG-SC Neurod2 Meis2 Glut', 'PVT-PT Ntrk1 Glut']\n",
      "p74, 139566, ['Astro-NT NN']\n",
      "p79, 94733, ['CBX MLI Cdh22 Gaba', 'Microglia NN']\n",
      "p8, 29448, ['AAA-BST Six3 Meis2 Gaba', 'IT AON-TT-DP Glut', 'MEA Slc17a7 Glut', 'MEA-BST Sox6 Gaba', 'MH Tac2 Glut']\n",
      "p9, 29845, ['BST MPN Six3 Gaba', 'Car3 Glut', 'HPF CR Glut', 'MM-ant Foxb1 Glut', 'MY Prox1 Gly-Gaba', 'SCop Sln Glut', 'SCsg Gabrr2 Gaba', 'SCsg Pde5a Glut', 'SPVI-SPVC Mafa Glut', 'TRS Sln Glut']\n"
     ]
    }
   ],
   "source": [
    "seq_annotation_df# Split the sequencing data for the input of the final round of integration.\n",
    "# Because imputation is done at this round, all genes are included.\n",
    "partitions = adata_seq.obs['integration_partition'].cat.categories\n",
    "\n",
    "print('Partition, N_cells, subclasses')\n",
    "for pn in partitions:\n",
    "    adata_subset = adata_seq[adata_seq.obs['integration_partition'] == pn]\n",
    "    p = os.path.join(partition_path, pn.replace('/', '-').replace(' ', '_'))\n",
    "    os.makedirs(p, exist_ok=True)\n",
    "    adata_subset.write_h5ad(os.path.join(p, 'adata_seq.h5ad'), compression='gzip')\n",
    "    \n",
    "    print(f'{pn}, {adata_subset.shape[0]}, {list(np.unique(adata_subset.obs[\"subclass_label\"]))}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a258853c-4997-4ac7-9029-73a86a913928",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata_seq, adata_merfish"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "188d46cd-3de7-4a65-9998-0e4a2faefb63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p40\n",
      "p28\n",
      "p56\n",
      "p45\n",
      "p79\n",
      "p15\n",
      "p26\n",
      "p44\n",
      "p4\n",
      "p8\n",
      "p6\n",
      "p48\n",
      "p5\n",
      "p41\n",
      "p33\n",
      "p3\n",
      "p51\n",
      "p20\n",
      "p13\n",
      "p10\n",
      "p2\n",
      "p36\n",
      "p54\n",
      "p37\n",
      "p22\n",
      "p34\n",
      "p17\n",
      "p38\n",
      "p27\n",
      "p53\n",
      "p9\n",
      "p42\n",
      "p12\n",
      "p16\n",
      "p18\n",
      "p7\n",
      "p39\n",
      "p21\n",
      "p11\n",
      "p23\n",
      "p24\n",
      "p46\n",
      "p19\n",
      "p74\n",
      "p1\n",
      "p0\n",
      "p14\n",
      "p68\n",
      "p55\n",
      "p59\n",
      "p50\n",
      "p25\n",
      "p29\n",
      "p69\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/doug.henze/.conda/envs/ABC_Download/lib/python3.9/site-packages/anndata/_core/anndata.py:1818: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata_seq_cg_list = []\n",
    "for pn in os.listdir(partition_path):\n",
    "    print(pn)\n",
    "    adata_seq = sc.read_h5ad(os.path.join(partition_path, pn, 'adata_seq.h5ad'))\n",
    "    adata_seq = adata_seq[:, adata_seq.var.index.isin(common_genes)].copy()\n",
    "    adata_seq_cg_list.append(adata_seq)\n",
    "    \n",
    "adata_seq_cg = anndata.concat(adata_seq_cg_list)\n",
    "adata_seq_cg.write_h5ad(os.path.join(workspace_path, 'adata_seq_common_genes.h5ad'), compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4f53dee2-ebc5-4a51-ae09-db20901b1f9c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ABC",
   "language": "python",
   "name": "abc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
